language/move-symbol-pool/src/symbol.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{pool::Entry, SYMBOL_POOL};
+use serde::{Deserialize, Serialize};
+use std::{cmp::Ordering, fmt, num::NonZeroU64};
+
+/// Represents a string that has been cached.
+///
+/// A `Symbol` represents a pointer to string data that is owned by the global
+/// symbol pool; it is not the string data itself. This enables this
+/// representation to implement `Copy` and other traits that some string types
+/// cannot.
+///
+/// The strings that `Symbol` types represent are added to the global cache as
+/// the `Symbol` are created.
+///
+/// ```
+///# use crate::move_symbol_pool::Symbol;
+/// let s1 = Symbol::from("hi"); // "hi" is stored in the global cache
+/// let s2 = Symbol::from("hi"); // "hi" is already stored, cache does not grow
+/// assert_eq!(s1, s2);
+/// ```
+///
+/// Use the method [`as_str()`] to access the string value that a `Symbol`
+/// represents. `Symbol` also implements the [`Display`] trait, so it can be
+/// printed as an ordinary string would. This makes it easier to use with
+/// crates that print strings to a terminal, such as codespan.
+///
+/// ```
+///# use crate::move_symbol_pool::Symbol;
+/// let message = format!("{} {}",
+///     Symbol::from("hello").as_str(),
+///     Symbol::from("world"));
+/// assert_eq!(message, "hello world");
+/// ```
+///
+/// [`as_str()`]: crate::move_symbol_pool::Symbol::as_str
+/// [`Display`]: std::fmt::Display
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct Symbol(NonZeroU64);
```

Comment:
To improve ergonomics of using `Symbol` you could look into implementing `AsRef<str>`

Reply:
Thanks! I added an implementation. I was bummed that it didn't allow me to remove any calls to `as_str` in #8742, though. I guess -- and since I'm still learning Rust hopefully you can teach me here -- that for that to have been possible, functions that take `&str` parameters could be rewritten to take `AsRef<str>` instead, where appropriate? I'll try to keep that in mind when writing my own functions from now on.

Reply:
Yep exactly that. There are tradeoffs* for making a function generic on `T: AsRef<str>` vs `&str` so i'm not saying you should prefer one over the other. Though it is beneficial to include the `AsRef` impl for the instances where you do have an API that takes in a `T: AsRef<str>`

\* Having lots of generic functions can increase compilation times and codesize since all instances are monomorphized. There are tricks to limiting that though, one such trick is having one API that takes in `T: AsRef<str>` where all its purpose is to call `as_ref()` to get the `&str` and then forward to another API that isn't generic and only takes `&str`s.

Reply:
I had thought it would be bad to do a bunch of Rust traityness as it tends to hide behavior. But maybe it doesn't really matter here, and is best to be more ergonomic.